File: app/src/main/java/fudi/freddy/permissionservice/MainActivity.java
package fudi.freddy.permissionservice;

import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;

import fudi.freddy.permission.PermissionService;

import static android.Manifest.permission.ACCESS_FINE_LOCATION;
import static android.Manifest.permission.CAMERA;

/**
 *
 * @author Romell Domínguez
 * @version 1.0.a 23/02/2017
 * @since 1.0
 */

public class MainActivity extends AppCompatActivity {

    TextView textView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if (savedInstanceState == null)
            initViews();
        new PermissionService(this).request(
                new String[]{ACCESS_FINE_LOCATION, CAMERA},
                callback);
    }

    private void initViews() {
        textView = (TextView) findViewById(R.id.text);
    }

    private void initController() {
        textView.setText("ToDo");
        Toast.makeText(getBaseContext(), textView.getText().toString(), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        // In case have other permissions flow task
//        if (requestCode == PermissionService.requestCode) {
        callback.handler(permissions, grantResults);
//        }
    }

    private PermissionService.Callback callback = new PermissionService.Callback() {
        @Override
        public void onRefuse(ArrayList<String> RefusePermissions) {
            Toast.makeText(getBaseContext(),
                    "Have to allow all permissions",
                    Toast.LENGTH_SHORT).show();

            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    finish();
                }
            }, 2000);
        }

        @Override
        public void onFinally() {
            initController();
        }
    };
}

File: app-kotlin/src/main/java/com/romellfudi/permissionservice/ui/main/view/MainActivity.kt
package com.romellfudi.permissionservice.ui.main.view

import android.app.Activity
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.view.View
import com.romellfudi.permission.PermissionService
import com.romellfudi.permissionservice.R
import com.romellfudi.permissionservice.ui.main.ActivityModule
import com.romellfudi.permissionservice.ui.main.component.ActivityComponent
import com.romellfudi.permissionservice.ui.main.component.DaggerActivityComponent
import com.romellfudi.permissionservice.ui.main.presenter.MainContract
import kotlinx.android.synthetic.main.activity_main.*
import javax.inject.Inject

/**
 *
 * @author Romell Domínguez
 * @version 1.0.a 23/02/2017
 * @since 1.0
 */

class MainActivity : AppCompatActivity(), MainContract.MainView {

    override val view: View
        get() = findViewById(android.R.id.content)

    @Inject
    lateinit var callback: PermissionService.Callback

    @Inject
    lateinit var permissionService: PermissionService

    private val activityComponent: ActivityComponent by lazy {
        DaggerActivityComponent.builder()
                .activityModule(ActivityModule(this))
                .build()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        activityComponent.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        askBtn.setOnClickListener { permissionService.request(this,callback) }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) =
            PermissionService.handler(callback, grantResults, permissions)

    override fun showOK() =
            Snackbar.make(view, "ready", Snackbar.LENGTH_SHORT).run {
                view.setBackgroundColor(
                        ContextCompat.getColor(this@MainActivity, R.color.colorPrimary))
                show()
            }

    override fun showError(error: String) =
            Snackbar.make(view, error, Snackbar.LENGTH_SHORT).run {
                view.setBackgroundColor(
                        ContextCompat.getColor(this@MainActivity, R.color.colorAccent))
                show()
            }
}
